ugs-core/test/com/willwinder/universalgcodesender/firmware/GrblFirmwareSettingsTest.java

```diff
@@ -227,4 +227,137 @@ public void setValueShouldTimeoutIfNoResponseFromController() throws Exception {
         verify(controller, times(1)).sendCommandImmediately(any());
         verify(firmwareSettingsListener, times(0)).onUpdatedFirmwareSetting(any());
     }
+
+    @Test
+    public void getInvertDirectionShouldReturnEachBitAsAxis() {
+        target.rawResponseListener("$3=0");
+        assertEquals(false, target.isInvertDirectionX());
+        assertEquals(false, target.isInvertDirectionY());
+        assertEquals(false, target.isInvertDirectionZ());
+
+        target.rawResponseListener("$3=1");
+        assertEquals(true, target.isInvertDirectionX());
+        assertEquals(false, target.isInvertDirectionY());
+        assertEquals(false, target.isInvertDirectionZ());
+
+        target.rawResponseListener("$3=2");
+        assertEquals(false, target.isInvertDirectionX());
+        assertEquals(true, target.isInvertDirectionY());
+        assertEquals(false, target.isInvertDirectionZ());
+
+        target.rawResponseListener("$3=4");
+        assertEquals(false, target.isInvertDirectionX());
+        assertEquals(false, target.isInvertDirectionY());
+        assertEquals(true, target.isInvertDirectionZ());
+
+        target.rawResponseListener("$3=7");
+        assertEquals(true, target.isInvertDirectionX());
+        assertEquals(true, target.isInvertDirectionY());
+        assertEquals(true, target.isInvertDirectionZ());
+    }
+
+    @Test
+    public void setInvertDirectionXShouldSetBit() throws FirmwareSettingsException, InterruptedException {
```

Comment:
I'm having a hard time following some of these. At one point I used `GUIHelpers.invokeLater` to optionally run things on the same thread to make unit testing a little easier. Do you think something like that could be done here? Perhaps move that function to `ThreadHelpers`.

Another thing I've done in other Controller/Communicator classes is to have a constructor which injects the communicator so that I can mock that piece for the validations.

Reply:
Yes, that's a bit cleaner, thanks!